tools/userland-fetch

```diff
@@ -28,246 +28,371 @@
+        if opt in ["-a", "--user-agent"]:
+            user_agent_arg = arg
+        elif opt in ["-f", "--file"]:
+            file_arg = arg
+        elif opt in ["-l", "--link"]:
+            link_arg = True
+        elif opt in ["-k", "--keep"]:
+            keep_arg = True
+        elif opt in ["-h", "--hash"]:
+            hash_arg = arg
+        elif opt in ["-s", "--search"]:
+            search_list.append(arg)
+        elif opt in ["-S", "--sigurl"]:
+        	sig_arg = arg
+        elif opt in ["-u", "--url"]:
+            url_arg = arg
+        else:
+            assert False, "unknown option"
+
+    if url_arg == None:
+        usage()
+
+    for name in download_from_paths(search_list, file_arg, url_arg, link_arg):
+        print "\n    validating signature...",
+
+        sig_valid = False
+        if not sig_arg:
+            print "skipping (no signature URL)"
+        else:
+            # Put the signature file in the same directory as the
+            # file we're downloading.
+            sig_file = os.path.join(
+                os.path.dirname(file_arg),
+                os.path.basename(sig_arg))
+            # Validate with the first signature we find.
+            for sig_file in download_from_paths(search_list, sig_file,
+                                                sig_arg, link_arg, True):
+                if sig_file:
+                    if validate_signature(name, sig_file):
+                        print "ok"
+                        sig_valid = True
+                    else:
+                        print "failed"
+                    break
+                else:
+                    continue
+            else:
+                print "failed (couldn't fetch signature)"
+
+        print "    validating hash...",
+        realhash = validate_container(name, hash_arg)
+
+        if not hash_arg:
+            print "skipping (no hash)"
+            print "hash is: %s" % realhash
+        elif realhash == hash_arg:
+            print "ok"
+        else:
+            payloadhash = validate_payload(name, hash_arg)
+            if payloadhash == hash_arg:
+                print "ok"
+            else:
+                # If the signature validated, then we assume
+                # that the expected hash is just a typo, but we
+                # warn just in case.
+                if sig_valid:
+                    print "invalid hash!"
+                else:
+                    print "corruption detected"
+
+                print "    expected: %s" % hash_arg
+                print "    actual:   %s" % realhash
+                print "    payload:  %s" % payloadhash
+
+                # An invalid hash shouldn't cause us to remove
+                # the target file if the signature was valid.
+                if not sig_valid:
+                    try:
+                        os.remove(name)
+                    except OSError:
+                        pass
+
+                    continue
+
+        sys.exit(0)
```

Comment:
what if there's no signature and hash invalid? Will this line be reached (hard to follow intendation :)